### PR TITLE
Add std::optional header-only library

### PR DESCRIPTION
**Pull request by alexdewar**
std::optional is a rather handy interface that is sadly only available from C++17 onwards.

I'm working on some code where it would make sense to use this kind of interface and I feel like it could be useful elsewhere.

### CODE REVIEW

File: third_party/optional.hpp [hunk 1]
```diff
@@ -0,0 +1,1076 @@
...
+# undef TR2_OPTIONAL_REQUIRES
+# undef TR2_OPTIONAL_ASSERTED_EXPRESSION
+
+/*
+ * Import everything into BoBRobotics namespace.
+ *
+ * This is a bit of a hack, but prefixing everything with std::experimental is
+ * annoyingly verbose.
+ */
+namespace BoBRobotics {
```
Comment (neworderofjamie) [inline third_party/optional.hpp:1072]: std::optional is definitely one of the most obviously useful bits of C++17! I kinda thing it should be up to the code that uses this to tersify ``std::experimental`` though

Comment (alexdewar) [reply third_party/optional.hpp:1072]: Hmm... You're probably right. I'll leave out the namespace malarkey then

Comment (alexdewar) [reply third_party/optional.hpp]: Are you happy to merge otherwise?

Comment (neworderofjamie) [reply third_party/optional.hpp]: yep! the nuance of a third-party backport of ``std::optional`` is beyond me 😄

Comment (alexdewar) [reply third_party/optional.hpp]: Haha - me too!

**Pull request merged**